archive/imo/imo2021_q1.lean

```diff
@@ -0,0 +1,246 @@
+import tactic data.real.sqrt
+
+/-
+# IMO 2021 Q1
+Let n≥100 be an integer. Ivan writes the numbers n, n+1,..., 2n each on different cards.
+He then shuffles these n+1 cards, and divides them into two piles. Prove that at least one
+of the piles contains two cards such that the sum of their numbers is a perfect square.
+# Solution
+We show there exists a triplet a, b, c ∈ [n , 2n] with a < b < c and each of the sums (a + b),
+(b + c), (a + c) being a perfect square. Specifically, we consider the linear system of equations
+
+  a + b = (2 * l - 1) ^ 2
+  a + c = (2 * l) ^ 2
+  b + c = (2 * l + 1) ^ 2
+
+which can be solved to give
+
+  a = 2 * l * l - 4 * l
+  b = 2 * l * l + 1
+  c = 2 * l * l + 4 * l
+
+Therefore, it is enough to show that there exists a natural number l such that
+n ≤ 2 * l * l - 4 * l and 2 * l * l + 4 * l ≤ 2 * n for n ≥ 100.
+
+Then, by
+Pigeonhole principle, at least two numbers in the triplet must lie in the same pile, which
+finishes the proof.
+-/
+
+
+open real
+
+lemma lower_bound (n l : ℕ) (hl : 2 + sqrt (4 + 2 * n) ≤ 2 * (l : ℝ)) :
+  (n : ℕ) + 4 * l ≤ 2 * l * l :=
+begin
+  have h₁ : sqrt (4 + 2 * n) ≤ 2 * l - 2 := by linarith,
+  replace h₁ := (sqrt_le_iff.1 h₁).2,
+  ring_exp at h₁, norm_num at h₁, norm_cast at h₁,
+  linarith,
+end
+
+lemma upper_bound (n l : ℕ) (hl₂ : (l : ℝ) ≤ sqrt (1 + n) - 1):
+2 * l * l + 4 * l ≤ 2 * n :=
+begin
+  have h₁ : (l : ℝ) + 1 ≤ sqrt (1 + n) := by linarith,
+  rw le_sqrt at h₁,
+  ring_exp at h₁, norm_num at h₁,
+  norm_cast at h₁,
+  linarith,
+  all_goals {norm_cast, linarith},
+end
+
+
+lemma radical_inequality {n : ℕ} (h : n ≥ 107) : sqrt (4 + 2 * n) ≤ 2 * (sqrt (1 + n) - 3) :=
+begin
+  rw sqrt_le_iff,
+  split,
+  { norm_num,
+    rw le_sqrt,
+    all_goals {norm_cast, linarith}},
+  ring_exp,
+  rw [pow_two, ← sqrt_mul, sqrt_mul_self],
+  suffices : 24 * sqrt (1 + n) ≤ 2 * n + 36,
+  linarith,
+  rw mul_self_le_mul_self_iff,
+  ring_exp,
+  rw [pow_two, ← sqrt_mul, sqrt_mul_self],
+  --Not splitting into cases lead to a deterministic timeout on my machine
+  by_cases p: n ≥ 108,
+  norm_cast,
+  nlinarith,
+  have : n = 107 := by linarith,
+  subst this,
+  norm_num,
+  swap 3, norm_num, apply sqrt_nonneg,
+  all_goals {norm_cast, linarith},
+end
+
+/-We will later make use of the fact that there exists (l : ℕ) such that
+n ≤ 2 * l * l - 4 * l and 2 * l * l + 4 * l ≤ 2 * n for n ≥ 107.
+-/
+lemma exists_numbers_in_interval (n : ℕ) (hn : n ≥ 107): ∃ (l : ℕ),
+(n + 4 * l ≤ 2 * l * l ∧ 2 * l * l + 4 * l ≤ 2 * n) :=
+begin
+  suffices : ∃ (l : ℕ), 2 * (l : ℝ) ≥ 2 + sqrt (4 + 2 * n) ∧ (l : ℝ) ≤ sqrt (1 + n) - 1,
+  cases this with l t,
+  use l,
+  refine ⟨lower_bound n l t.1, upper_bound n l t.2⟩,
+  lift floor (sqrt (1 + n) - 1) to ℕ with x,
```

Comment:
hmm; you could also use `nat.sqrt` instead of this `real.sqrt`, which is automatically floored; obviously this is another rewrite though :( there's also `nat_floor`, but I don't super terribly awfully mind this